### DISCUSSION TITLE
ffmpeg in the ImageMagick directory

### DISCUSSION THREAD

**Discussion by cm2187** [ImageMagick/ImageMagick#3131] [Development]
I don't want to make it an issue because it is kind of a trivial problem. But ImageMagick is installed with ffmpeg.exe in its root directory and adds that directory to the environment path variable in windows. In my case I ended up with the imagemagick folder at the top of my paths, which means that when I typed "ffmpeg.exe" from any console, it is the version from the ImageMagick folder that ran instead of the default one. Ffmpeg introduced a breaking change recently which is why I realised this (the version from the ImageMagick folder is slightly older). 

It's kind of a trivial thing to work around, you just need to reorder your path environment variables. But I wonder if ImageMagick shouldn't push ffmpeg to a sub directory to avoid overwriting the default ffmpeg to be used on the system. As it may break silently people's workflows by making them use an older or more recent version of ffmpeg they don't expect.

**Comment by dlemstra** [comment]
The Windows installers has an option to disable the installation of `FFmpeg`. You should set that if you would want to use your own version of `FFMpeg`.

**Comment by cm2187** [comment]
I suspect this does something different. I don't want to change the version of ffmpeg that Imagemagick is using (it is better to let ImageMagick use the version it was tested for and therefore install its own version in its own directory). I would just prefer that ImageMagick doesn't overrides the version of ffmpeg used by default when you type ffmpeg in any command line. For instance by moving its own ffmpeg to a sub directory so it is not included in default paths.

Again not a big deal for me. It is just that I can think of how it could break some workflows without people realising it. So thought I would mention it as a suggestion.

**Comment by gaiking-uk** [comment]
### TLDR
How do I configure the ffmpeg path for ImageMagick?

### Additional Info
I saw the options for ```install ffmpeg``` and ```add to PATH``` during the setup, and as I already have FFmpeg installed (and used/referenced by multiple other apps) I didn't want to install it again (or modify PATH) so I unselected these options during install. - - - **However, now that ImageMagick is installed, how do I *tell* it where it can find the FFmpeg binaries?**

NB: Failing the existence of something like a ```FFmpeg_path``` parameter in ```configure.xml``` I would normally create a junction for the FFmpeg-using app however in this case I don't know where ImageMagick would expect this to be... in the root, or in modules, etc(?)

**Comment by snibgo** [comment]
By experiment (on Windows 11), I find that IM will use the ffmpeg.exe from the directory that magick.exe is in, even if that directory is not on the system path, provided the directory contains ffmpeg.exe. Otherwise, IM will search the system path.

IM has no special facility to declare that ffmpeg is in some directory that is not on the path. However, I expect that editing the entries in delegates.xml would do that.

Does that answer the question?

**Reply by GeeMack** [reply]
For several years when I've installed updates of IM, I chose to not have it install ffmpeg. Then whenever I update ffmpeg, I just put a copy of the EXE in the IM folder, overwriting the previous version. IM apparently finds it in the IM folder, and since that's the one in the PATH, the same ffmpeg runs from the command line. I've never encountered any issues that I'm aware of with either.

**Comment by gaiking-uk** [comment]
Thanks @snibgo @GeeMack for the quick reply and info! 👍🏼

I don't currently have FFmpeg in the system ```%PATH%``` however that sounds like the quick / easiest option and I don't mind adding a *limited* number of program-specific locations to the path (I do try to keep it tidy though as nowadays ***sooo*** many installs try to add to it!). Alternatively (more as an FYI to future readers as will probably go with the above option) -- if IM expects to find ```FFmpeg.exe``` in the IM root folder, then I could create a symbolic/hard link to the FFmpeg.exe binary's source location.

Cool, great to get it sorted straight away, thanks again!

***

**UPDATE:**   I hadn't got around to checking ```Delegates.xml``` when I posted (I was considering switching to the portable version of IM, but I see this includes a delegates.xml file also).

Building on what @snibgo alluded to, there are 2 lines in ```delegates.xmll``` that specify the path IM should use for encoding and decoding using FFmpeg...

```
<delegate decode="video:decode" command="&quot;ffmpeg.exe&quot; -nostdin -loglevel error -i &quot;%s&quot; -an -f rawvideo -y %s &quot;%s&quot;"/>
<delegate encode="video:encode" stealth="True" command="&quot;ffmpeg.exe&quot; -nostdin -loglevel error -i &quot;%s%%d.%s&quot; %s &quot;%s.%s&quot;"/>
```

So looks like you can configure FFmpeg here by replacing ```&quot;ffmpeg.exe&quot;``` with the path to FFmpeg.exe 👍🏼
